test/e2e/storage/drivers/proxy/portproxy.go

```diff
@@ -0,0 +1,330 @@
+	ctx, cancel := context.WithCancel(ctx)
+	l := &listener{
+		connections: make(chan *connection),
+		ctx:         ctx,
+		cancel:      cancel,
+		addr:        addr,
+	}
+
+	var connectionsCreated, connectionsClosed int32
+
+	runForwarding := func() {
+		klog.V(2).Infof("%s: starting connection polling", prefix)
+		defer klog.V(2).Infof("%s: connection polling ended", prefix)
+
+		// This delay determines how quickly we notice when someone has
+		// connected inside the cluster. With socat, we cannot make this too small
+		// because otherwise we get many rejected connections. With the mock
+		// driver as proxy that doesn't happen as long as we don't
+		// ask for too many concurrent connections because the mock driver
+		// keeps the listening port open at all times and the Linux
+		// kernel automatically accepts our connection requests.
+		tryConnect := time.NewTicker(100 * time.Millisecond)
+		defer tryConnect.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tryConnect.C:
+				currentClosed := atomic.LoadInt32(&connectionsClosed)
+				openConnections := connectionsCreated - currentClosed
+				if openConnections >= maxConcurrentConnections {
+					break
+				}
+
+				klog.V(5).Infof("%s: trying to create a new connection #%d, %d open", prefix, connectionsCreated, openConnections)
+				stream, err := dial(ctx, fmt.Sprintf("%s #%d", prefix, connectionsCreated), dialer, addr.Port)
+				if err != nil {
+					klog.Errorf("%s: no connection: %v", prefix, err)
+					break
+				}
+				// Make the connection available to Accept below.
+				klog.V(5).Infof("%s: created a new connection #%d", prefix, connectionsCreated)
+				l.connections <- &connection{
+					stream:  stream,
+					addr:    addr,
+					counter: connectionsCreated,
+					closed:  &connectionsClosed,
+				}
+				connectionsCreated++
+			}
+		}
+	}
+
+	// Portforwarding and polling for connections run in the background.
+	go func() {
+		for {
+			running := false
+			pod, err := clientset.CoreV1().Pods(addr.Namespace).Get(ctx, addr.PodName, metav1.GetOptions{})
+			if err != nil {
+				klog.V(5).Infof("checking for container %q in pod %s/%s: %v", addr.ContainerName, addr.Namespace, addr.PodName, err)
+			}
+			for i, status := range pod.Status.ContainerStatuses {
+				if pod.Spec.Containers[i].Name == addr.ContainerName &&
+					status.State.Running != nil {
+					running = true
+					break
+				}
+			}
+
+			if running {
+				klog.V(2).Infof("container %q in pod %s/%s is running", addr.ContainerName, addr.Namespace, addr.PodName)
+				runForwarding()
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			// Sleep a bit before restarting. This is
+			// where we potentially wait for the pod to
+			// start.
+			case <-time.After(1 * time.Second):
+			}
+		}
+	}()
+
+	return l, nil
+}
+
+// Addr contains all relevant parameters for a certain port in a pod.
+// The container must be running before connections are attempted.
+type Addr struct {
```

Comment:
It looks like it implements some interface, please add a check for it.

Reply:
net.Addr, added.